module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,386 @@
+        )
+
+        // Project Manager Preferences
+        private val ignoredPathsString = preferenceMgr.ignoredPaths.joinToString(",")
+        var ignoredPaths = PreferenceField.TextInput(
+            this, initialValue = ignoredPathsString,
+            label = PROJECT_IGNORED_PATHS, placeholder = IGNORED_PATHS_PLACEHOLDER
+        )
+
+        // Query Runner Preferences
+        var matchQueryLimit = PreferenceField.TextInput(
+            this, initialValue = preferenceMgr.matchQueryLimit.toString(),
+            label = SET_QUERY_LIMIT, placeholder = QUERY_LIMIT_PLACEHOLDER,
+            caption = Optional.of(PREFERENCES_MATCH_QUERY_LIMIT_CAPTION)
+        ) {/* validator = */ it.toLongOrNull() != null }
+        
+        // Text Editor Preferences
+        var autoSave = PreferenceField.Checkbox(
+            this, initialValue = preferenceMgr.autoSave, label = ENABLE_EDITOR_AUTOSAVE
+        )
+
+        val preferenceGroups = listOf(
+            PreferenceGroup(GRAPH_VISUALISER, preferences = listOf(graphOutput)),
+            PreferenceGroup(PROJECT_MANAGER, preferences = listOf(ignoredPaths)),
+            PreferenceGroup(QUERY_RUNNER, preferences = listOf(matchQueryLimit)),
+            PreferenceGroup(TEXT_EDITOR, preferences = listOf(autoSave)),
+        )
+
+        val rootPreferenceGroup = PreferenceGroup(entries = preferenceGroups)
+
+        override fun cancel() {
+            StudioState.preference.preferencesDialog.close()
+        }
+
+        fun apply() {
+            if (isValid()) {
+                trySubmit()
+                modified = false
+            }
+        }
+
+        fun ok() {
+            apply()
+            cancel()
+        }
+
+        override fun isValid(): Boolean {
+            return rootPreferenceGroup.isValid()
+        }
+
+
+        override fun trySubmit() {
+            preferenceMgr.autoSave = autoSave.value
+            preferenceMgr.ignoredPaths = ignoredPaths.value.split(',').map { it.trim() }
+            preferenceMgr.matchQueryLimit = matchQueryLimit.value.toLong()
+            preferenceMgr.graphOutputEnabled = graphOutput.value
+        }
+    }
+
+    class PreferenceGroup(
+        override val name: String = "",
+        override var entries: List<PreferenceGroup> = emptyList(),
+        var preferences: List<PreferenceField> = emptyList(),
+        val content: @Composable () -> Unit = {},
+    ) : Navigable<PreferenceGroup> {
+
+        override var parent: Navigable<PreferenceGroup>? = null
+        override val info: String? = null
+        override val isExpandable = entries.isNotEmpty()
+        override val isBulkExpandable = false
+
+        override fun reloadEntries() {}
+
+        override fun compareTo(other: Navigable<PreferenceGroup>): Int {
+            return this.name.compareTo(other.name);
+        }
+
+        fun isValid(): Boolean {
+            return preferences.fold(true) { acc, preferenceField -> acc && preferenceField.isValid() } &&
```

Comment:
Oh, this takes me back to the good old Haskell days.

In seriousness, this is a terribly fancy way of writing `preferences.all { it.isValid() }`.

Reply:
I forget Kotlin has everything and the kitchen sink in terms of convenience. Done.